SSA/Projects/InstCombine/LLVM/Semantics.lean

```diff
@@ -78,11 +78,25 @@ def add? {w : Nat} (x y : BitVec w) : IntW w :=
 theorem add?_eq : LLVM.add? a b  = .some (BitVec.add a b) := rfl
 
 @[simp_llvm_option]
-def add {w : Nat} (x y : IntW w) : IntW w := do
+def add {w : Nat} (x y : IntW w) (nswnuw : Bool × Bool := ⟨ false , false ⟩ ) : IntW w := do
   let x' ← x
   let y' ← y
-  add? x' y'
-
+  let nsw := nswnuw.1
+  let nuw := nswnuw.2
+  if (nsw ∧  (x'.toInt + y'.toInt) < -(2^(w-1)) ∧ (x'.toInt + y'.toInt) ≥ 2^w) ∨ (nuw ∧  (x'.toNat + y'.toNat) ≥ 2^w) then
+    none
+  else
+    add? x' y'
+
+set_option allowUnsafeReducibility true
+@[simp, reducible]
+theorem add_reduce (x y : IntW w) :  add x y = match x , y with
+  | .none , _ => .none
+  | _ , .none => none
+  | .some a , .some b  => .some (a + b) := by
+  rcases x
+  all_goals (cases y)
+  all_goals (try simp ; try rfl)
```

Comment:
Correct me if I am wrong, but I was under the impression that this can be replaced by: 
```suggestion
  rcases x <;> cases y <;> [simp, rfl]
 ```

Reply:
Also, `rcases` without a pattern seems like it should be just a regular `cases`, no?